readme.md
# SCC Convertor for PollyNET Products
![Build](https://github.com/PollyNET/polly2scc/workflows/Build/badge.svg?branch=master)

At June 24, 2019, the new EARLINET database was released and new data format for the netCDF files was applied. Therefore, in order to be compatible with this update, we (PollyNET group) need to change the output results from Picasso (PollyNET automatic processing program) to be adapted to this change.

## Requirements

- python 3.5
- [`Anaconda3`][1] (Python 3.x version)

## Installation

### python virtual environment and install python dependencies

If you have installed [`Anaconda3`][1], you can easily setup the python environment for running the program. Otherwise, you can go [here][2] for some details of the installation.

**create a new virtual environment.**

```bash
conda create -n polly_earlinet
```

**activate the virtual environment**

```bash
activate polly_earlinet # windows
source activate polly_earlinet # linux
```

**install python3.5**

```bash
conda install python=3.5
```

**build the repository**

```bash
pip install git+https://github.com/PollyNET/polly2scc.git
```

## Usage

After the steps mentioned above, the command line tool **polly2scc** should be ready in your system **PATH**. You can see the instructions of the command with using `polly2scc -h`

```text
usage: polly2scc [-h] [-p POLLY_TYPE] [-l LOCATION] [-t FILE_TYPE] [-c CATEGORY]
              [-f FILENAME] [-d OUTPUT_DIR]
              [--range_e RANGE_LIM_E RANGE_LIM_E]
              [--range_b RANGE_LIM_B RANGE_LIM_B] [--camp_info CAMP_INFO]
              [--force]
              {list} ...

convert the polly profiles from labview program to EARLINET format

positional arguments:
  {list}                list supported campaign and instruments.
    list                list supported campaign and instruments.

optional arguments:
  -h, --help            show this help message and exit
  -p POLLY_TYPE, --polly_type POLLY_TYPE
                        setup the instrument type
  -l LOCATION, --location LOCATION
                        setup the campaign location
  -t FILE_TYPE, --file_type FILE_TYPE
                        setup the type of the profile (labview | picasso)
  -c CATEGORY, --category CATEGORY
                        setup the category of the profile (user_defined_category)
                        flag_masks: 1, 2, 4, 8, 16, 32, 64, 128, 256, 512
                        flag_meanings: cirrus climatol dicycles etna forfires
                        photosmog rurban sahadust stratos satellite_overpasses
  -f FILENAME, --filename FILENAME
                        setup the filename of the polly profile
  -d OUTPUT_DIR, --output_dir OUTPUT_DIR
                        setup the directory for the converted files
  --range_e RANGE_LIM_E RANGE_LIM_E
                        setup the height range for the converted e-files.
                        (e.g., --range_e 200 16000)
  --range_b RANGE_LIM_B RANGE_LIM_B
                        setup the height range for the converted b-files.
                        (e.g., --range_b 200 16000)
  --camp_info CAMP_INFO
                        setup the campaign info file [*.toml].
                        If not set, the program will search the config folder for a suitable one.
  --force               whether to overwrite the nc files if they exists
```

**Display the supported polly types**

```bash
polly2scc list --instrument
```

**Display the supported campaigns**

```bash
polly2scc list --campaign
polly2scc list --all   # show all campaigns together with the systems
```

**Convert one file**

```bash
polly2scc -p pollyxt_lacros -l punta_arenas -t labview -c 2 -f /User/zhenping/desktop/file1.txt -d /Users/zhenping/Destkop/test --force
```

|output file namings|example|description|
|:----------:|:-----:|:----------|
|{date:yyyymmdd\_HH:MM}\_{smooth:03d}\_{station ID}\_{PollyType}\_b355.nc|20190723\_1900\_075\_lei\_arielle\_b355.nc|results associated with backscatter, vol/par depolarization ratio at 355 nm|
|{date:yyyymmdd\_HH:MM}\_{smooth:03d}\_{station ID}\_{PollyType}\_b532.nc|20190723\_1900\_075\_lei\_arielle\_b532.nc|results associated with backscatter, vol/par depolarization ratio at 532 nm|
|{date:yyyymmdd\_HH:MM}\_{smooth:03d}\_{station ID}\_{PollyType}\_b1064.nc|20190723\_1900\_075\_lei\_arielle\_b1064.nc|results associated with backscatter coefficient at 1064 nm|
|{date:yyyymmdd\_HH:MM}\_{smooth:03d}\_{station ID}\_{PollyType}\_e355.nc|20190723\_1900\_075\_lei\_arielle\_e355.nc|results associated with backscatter and extinction coefficients at 355 nm|
|{date:yyyymmdd\_HH:MM}\_{smooth:03d}\_{station ID}\_{PollyType}\_e532.nc|20190723\_1900\_075\_lei\_arielle\_e532.nc|results associated with backscatter and extinction coefficients at 532 nm|

**convert files with using wildcards**

```bash
polly2scc -p pollyxt_lacros -l punta_arenas -t labview -c 2 -f /User/zhenping/desktop/file*.txt -d /Users/zhenping/Destkop/test --force
```

## Q&A

If you have any questions, please go to the [`issues`][4] session to check whether there was an answer. If not, please contact [me](#contact) or draft a new issue there.

## Acknowledgement

Special thanks to the persistent help from [Holger Baars][5] to test uploading the converted files to the new EARLINET database, which helps a lot for the debugging. In addition, I thank [Martin Radenz](radenz@tropos.de) to solve the file encoding issue.

> The repository is dependent on the [lidar_molecular][3], developed by **Iannis Binietoglou**. Thanks for sharing your great work.

## License

MIT License

Feel free to distribute it!!! :beer::beer::beer:

## Contact

Zhenping Yin 
<zhenping@tropos.de>

[1]: https://www.anaconda.com/distribution/
[2]: https://github.com/ZPYin/Pollynet_Processing_Chain/blob/master/doc/anaconda_installation.md
[3]: https://bitbucket.org/iannis_b/lidar_molecular/
[4]: https://github.com/ZPYin/polly_2_earlinet_convertor/issues
[5]: https://www.tropos.de/institut/ueber-uns/mitarbeitende/holger-baars

CHANGELOG
CHANGES
=======

* convert Picasso output p2e\_go to earlinet data format using cloud\_info.nc
* docs: change LICENSE
* docs: update license
* style: add help msg for error
* fix(polly2scc.py): crop ext range
* docs: update the introduction doc
* docs: update program\_diagram
* docs(changelog): update chagnelog
* docs: fix merge conflicts

v1.3
----

* feat: rename the repo as polly2scc
* fix: correct temperature with Kelvin
* feat: rename the repo as polly2scc

v1.2
----

* style: test pip upgrade
* Package managing with pbr
* refactor: rm \_\_init\_\_.py
* improvement(setup.py): manage python packaging with pbr
* improvement: replace the deprecated usage of pip
* style: remove unnecessary files
* improvement: enable output version
* fix: add missing input for list\_avail\_prodType
* Write processor\_version to EARLINET nc file
* improvement: write processor\_version
* test: replace test data file
* improvement: check picasso version before conversion
* docs: update program diagram
* docs(readme): update build badge
* refactor(setup.py): rename the module
* ci(build.yml): remove macOS and keep only py3.5
* ci(build.yml): upgrade pip
* docs(readme.md): update the badge
* style(build.yml): test github action
* Update github actions
* docs(readme.md): add CI badge
* ci(buiid.yml): add github actions
* docs: add introduction doc

v1.1
----

* docs(polly\_2\_earlinet\_convertor.py): fix typo
* feat: convert picasso profile to earlinet
* build: add lidar\_molecular
* refactor: fix merge conflicts
* docs: implement dev\_log
* docs: implement dev\_log
* docs(changelog): update CHANGELOG
* docs(readme.md): update readme

v1.0
----

* style: version 1.0
* style: version 1.0
* Update readme.md
* Fix bug of timestamp conversion and height
* Add readme badges
* Improvement: turn on the compression
* Improvements: replace the vertical\_resolution of the range bins with the effective resolution, which was expained in Sect. 4.2 in I. Mattis, et al, ATM, 2016
* Update: update the readme.md
* Bugfixes: enable the input range to start from 0
* Update readme.md
* Improvements: add the smoothing window to filenames to differenciate results from different smoothing settings
* Improvements: individual settings for range limit of e file (e355, 3532) and b file (b355, b532, b1064)
* Improvements: add -r to control the range limit of the variables
* Bugfixes: update the readme.md with removing the typos
* Update: add two non-EARLINET stations
* Improvements: add error handles for argparse
* Updates: implement new HOI IDs
* Bugfixes: assign specified 'earlinet\_product\_type' to different nc files
* Improvements: add the comments for the function of p2e\_go
* Update: add the hoi\_ids for limassol
* Improvemetns: tiny update for the readme.md
* Improvements: throw warning if old labview results were input
* Improvements: update the items in campaign info files
* Bugfixes: replace the measurement\_time in the metadata with the start\_time and stop\_time of the averaged file
* Bugfixes: 1. station\_ID to lower case: LEI -> lei
* Improvements: add exemplified data files from old version labview program
* revise the readme.md
* Add TODO
* Update the readme
* Bugfixes: specify the encoding for loading the toml files
* delete the log file
* Beta version
* test
* Improvements: Add keyword to show the whole campaign list
* Improvements: complete the configs
* backup
* test
* backup
* revise the readme
* alpha version
* Backup from local
* Backup from local
* Add MIT license
* Auto backup


tests/test_polly_2_earlinet_convertor.py
import sys
import os
import unittest
import shutil

projectDir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
tmpDir = os.path.join(projectDir, 'data', 'tmp')
sys.path.append(os.path.join(projectDir, 'src'))

from polly2scc import *


class Test(unittest.TestCase):

    @classmethod
    def setUpClass(self):
        print('Start to test polly2scc...')
        if not os.path.exists(tmpDir):
            os.mkdir(tmpDir)

    @classmethod
    def tearDownClass(self):
        if os.path.exists(tmpDir):
            shutil.rmtree(tmpDir)

        print('Finish testing polly2scc!')

    def setUp(self):
        print("execute setUp")

    def tearDown(self):
        print("execute tearDown")

    def test_find_in_string(self):
        print('---> Test on find_in_string')

        inStr = 'Reference value: 1.000000e-02 [Mm^{-1}*Sr^{-1}]; Reference' +\
                ' search range:   500.00 - 14000.00 [m]; Smoothing window: ' +\
                '4.575000e+02 [m]; Angstroem exponent: 0.90'

        value = find_in_string((
                    r'(?<=Reference search range:......... - )\d+.\d+',
                    float,
                    0
                ), inStr)

        print(value)
        self.assertEqual(value, 14000)

    def test_show_list(self):
        print('---> Test show_list')

        p2eConvertor = polly_2_earlinet_convertor()
        camp_dict = p2eConvertor.campaign_dict

        self.assertGreater(len(camp_dict), 0)

    def test_search_data_files(self):
        print('---> Test on search_data_file')

        p2eConvertor = polly_2_earlinet_convertor('arielle', 'leipzig')
        p2eConvertor.fileType = 'labview'

        # test search labview files
        labviewFiles = p2eConvertor.search_data_files(
            'le_*smooth.txt', filepath=os.path.join(projectDir, 'data')
        )
        self.assertEqual(len(labviewFiles), 1)
        self.assertEqual(
            os.path.basename(labviewFiles[0]),
            'le_arielle-20190723_2100-0058-49smooth.txt')

        # test search picasso files
        picassoFiles = p2eConvertor.search_data_files(
            '*TROPOS*profiles.nc', filepath=os.path.join(projectDir, 'data')
        )
        self.assertEqual(len(picassoFiles), 1)
        self.assertEqual(
            os.path.basename(picassoFiles[0]),
            '2020_05_06_Wed_TROPOS_00_00_01_0000_0059_profiles.nc'
        )

    def test_convert_labview_file(self):
        print('---> Test on convert labview file')

        p2eConvertor = polly_2_earlinet_convertor(
            'arielle', 'leipzig',
            fileType='labview', category=1,
            output_dir=os.path.join(projectDir, 'data', 'tmp'),
            force=True)

        fileLists = p2eConvertor.search_data_files(
            'le_*smooth.txt', filepath=os.path.join(projectDir, 'data')
        )

        for task in fileLists:
            dims, data, global_attris = p2eConvertor.read_data_file(task)

            b355 = p2eConvertor.write_to_earlinet_nc(
                    data, dims, global_attris,
                    range_lim=[0, 15000], prodType='b355')
            self.assertTrue(os.path.exists(b355))
            os.remove(b355)

            e355 = p2eConvertor.write_to_earlinet_nc(
                    data, dims, global_attris,
                    range_lim=[0, 15000], prodType='e355')
            self.assertTrue(os.path.exists(e355))
            os.remove(e355)

            b532 = p2eConvertor.write_to_earlinet_nc(
                    data, dims, global_attris,
                    range_lim=[0, 15000], prodType='b532')
            self.assertTrue(os.path.exists(b532))
            os.remove(b532)

            e532 = p2eConvertor.write_to_earlinet_nc(
                    data, dims, global_attris,
                    range_lim=[0, 15000], prodType='e532')
            self.assertTrue(os.path.exists(e532))
            os.remove(e532)

            b1064 = p2eConvertor.write_to_earlinet_nc(
                    data, dims, global_attris,
                    range_lim=[0, 15000], prodType='b1064')
            self.assertTrue(os.path.exists(b1064))
            os.remove(b1064)

    def test_convert_picasso_file(self):
        print('---> Test on convert picasso file')

        p2eConvertor = polly_2_earlinet_convertor(
            'PollyXT_TROPOS', 'leipzig',
            fileType='picasso', category=1,
            output_dir=os.path.join(projectDir, 'data', 'tmp'),
            force=True)

        fileLists = p2eConvertor.search_data_files(
            '*TROPOS*profiles.nc', filepath=os.path.join(projectDir, 'data')
        )

        for task in fileLists:
            dims, data, global_attris = p2eConvertor.read_data_file(task)

            b355 = p2eConvertor.write_to_earlinet_nc(
                    data, dims, global_attris,
                    range_lim=[0, 15000], prodType='b355')
            self.assertTrue(os.path.exists(b355))
            os.remove(b355)

            e355 = p2eConvertor.write_to_earlinet_nc(
                    data, dims, global_attris,
                    range_lim=[0, 15000], prodType='e355')
            self.assertTrue(os.path.exists(e355))
            os.remove(e355)

            b532 = p2eConvertor.write_to_earlinet_nc(
                    data, dims, global_attris,
                    range_lim=[0, 15000], prodType='b532')
            self.assertTrue(os.path.exists(b532))
            os.remove(b532)

            e532 = p2eConvertor.write_to_earlinet_nc(
                    data, dims, global_attris,
                    range_lim=[0, 15000], prodType='e532')
            self.assertTrue(os.path.exists(e532))
            os.remove(e532)

            b1064 = p2eConvertor.write_to_earlinet_nc(
                    data, dims, global_attris,
                    range_lim=[0, 15000], prodType='b1064')
            self.assertTrue(os.path.exists(b1064))
            os.remove(b1064)

    def test_list_avail_prodType(self):
        print('---> Test on list_avail_prodType')

        p2eConvertor = polly_2_earlinet_convertor(
            'PollyXT_TROPOS', 'leipzig',
            fileType='picasso', category=1,
            output_dir=os.path.join(projectDir, 'data', 'tmp'),
            force=True)

        fileLists = p2eConvertor.search_data_files(
            '*TROPOS*profiles.nc', filepath=os.path.join(projectDir, 'data')
        )

        dims, data, global_attris = p2eConvertor.read_data_file(fileLists[0])
        availProds = p2eConvertor.list_avail_prodType(data)

        self.assertListEqual(
            availProds,
            ['b355', 'e355', 'b532', 'e532', 'b1064'])

    def test_polly2scc(self):
        print('---> Test on p2e_go')

        polly2scc(
            'PollyXT_TROPOS', 'leipzig', 'picasso', 1,
            os.path.join(
                projectDir, 'data',
                '2020_05_06_Wed_TROPOS_00_00_01_0000_0059_profiles.nc'),
            tmpDir,
            [0, 14000], [0, 15000], 'Leipzig_campaign_info_9.toml', True)

        self.assertTrue(os.path.exists(
            os.path.join(
                tmpDir, '20200506_0029_0458_lei_pollyxt_tropos_b355.nc')))
        self.assertTrue(os.path.exists(
            os.path.join(
                tmpDir, '20200506_0029_0458_lei_pollyxt_tropos_e355.nc')))
        self.assertTrue(os.path.exists(
            os.path.join(
                tmpDir, '20200506_0029_0458_lei_pollyxt_tropos_b532.nc')))
        self.assertTrue(os.path.exists(
            os.path.join(
                tmpDir, '20200506_0029_0458_lei_pollyxt_tropos_e532.nc')))
        self.assertTrue(os.path.exists(
            os.path.join(
                tmpDir, '20200506_0029_0458_lei_pollyxt_tropos_b1064.nc')))


def main():

    suite = unittest.TestSuite()

    tests = [
        Test('test_find_in_string'),
        Test('test_show_list'),
        Test('test_list_avail_prodType'),
        Test('test_search_data_files'),
        Test('test_convert_labview_file'),
        Test('test_convert_picasso_file')
        ]   # setup the test list
    suite.addTests(tests)

    # print the test report as html file
    # with open('HTMLReport.html', 'w') as f:
    #     runner = HTMLTestRunner(stream=f,
    #                             title='MathFunc Test Report',
    #                             description='generated by HTMLTestRunner.',
    #                             verbosity=2
    #                             )
    #     runner.run(suite)

    # print the test report to the console
    runner = unittest.TextTestRunner(verbosity=2)
    runner.run(suite)


if __name__ == '__main__':
    main()


include/iannis_b-lidar_molecular-de3d2ef2f36b/tests/test_rayleigh_scattering_bucholtz.py
"""
Unit test of the rayleigh_scattering_bucholtz functions.

The test is performed using the tabular values found in the referenced paper.

References
----------
Bucholtz, A. Rayleigh-scattering calculations for the terrestrial atmosphere.
Applied Optics, Vol. 3

.. todo::
   * Move deleted plots in a dedicated Notebook.
"""

import os
import unittest2 as unittest
from scipy.interpolate import interp1d
from molecular.rayleigh_scattering_bucholtz import *

# Get the data path
current_path = os.path.dirname(__file__)
data_base_path = os.path.join(current_path, '../data/bucholtz_tabular_values/')


class TestKingCorrectionFactor(unittest.TestCase):
    def test_against_bucholtz_paper_table_1(self):
        #   Specify the data file.
        filename_reference = os.path.join(data_base_path, "bucholtz_1995_table_1.csv")        
        
        #   Load the reference data.
        reference = np.loadtxt(filename_reference, delimiter='\t', skiprows=1)        
        
        #   Calculate the King correction factor.
        calculated = king_correction_factor(reference[:, 0])

        #   Round the values to the desired decimal.
        calculated = np.around(calculated, decimals=3)

        #   Test if equal to reference.
        np.testing.assert_array_equal(calculated, reference[:, 1])

    
class TestScatteringCrossSection(unittest.TestCase):
    def test_against_bucholtz_paper_table_2(self):
        #   Specify the data file.
        filename_reference = os.path.join(data_base_path, "bucholtz_1995_table_2.csv")
        
        #   Load the reference data.
        reference = np.loadtxt(filename_reference, delimiter='\t', skiprows=1)

        #   Calculate the Rayleigh scattering cross section. (cm^2)
        calculated = scattering_cross_section(reference[:, 0])

        #   Round the values to the desired decimal.
        calculated[0:7] = np.around(calculated[0:7], decimals=28)
        calculated[7:26] = np.around(calculated[7:26], decimals=29)
        calculated[26:60] = np.around(calculated[26:60], decimals=30)
        calculated[60:67] = np.around(calculated[60:67], decimals=31)
        calculated[67:75] = np.around(calculated[67:75], decimals=32)
        calculated[75:] = np.around(calculated[75:], decimals=33)
        
        #   Assign zero to the expected failing wavelengths.
        index_fail = 26
        reference[index_fail, 1] = 0
        calculated[index_fail] = 0
        
        #   Test if equal to reference.
        np.testing.assert_allclose(calculated, reference[:, 1], rtol=1e-4)

    @unittest.skip("Skip the expected failing wavelengths.")
    def test_failing_wavelengths(self):
        #   Specify the data file.
        filename_reference = os.path.join(data_base_path, "bucholtz_1995_table_2.csv")
        
        #   Load the reference data.
        reference = np.loadtxt(filename_reference, delimiter='\t', skiprows=1)
        
        #   Specify the failing wavelength index.
        index_fail = 26
        
        #   Calculate the Rayleigh scattering cross section. (cm^2)
        calculated = scattering_cross_section(reference[index_fail, 0])

        #   Round the values to the desired decimal.
        calculated = np.around(calculated, decimals=30)
        
        #   Test if equal to reference.
        np.testing.assert_allclose(calculated, reference[index_fail, 1], rtol=1e-4)
    
    def test_last_working_values(self):
        #   Specify the data file.
        filename_reference = os.path.join(data_base_path, "bucholtz_1995_last_working_values.csv")

        #   Load the data from the file.
        reference = np.loadtxt(filename_reference, delimiter='\t', skiprows=1)

        #   Calculate the Rayleigh scattering cross section. (cm^2)
        calculated = scattering_cross_section(reference[:, 0])

        # Test if equal to reference.
        np.testing.assert_allclose(calculated, reference[:, 1], rtol=1e-6)
        

class TestVolumeScatteringCoefficient(unittest.TestCase):
    def test_against_bucholtz_paper_table_2(self):
        #   Specify the data file.
        filename_reference = os.path.join(data_base_path, "bucholtz_1995_table_2.csv")

        #   Load the reference data.
        reference = np.loadtxt(filename_reference, delimiter='\t', skiprows=1)

        #   Calculate the total Rayleigh volume-scattering coefficient. (km^-1)
        calculated = volume_scattering_coefficient(reference[:, 0], P_s, T_s)

        # Round the values to the desired decimal.
        calculated[0:13] = np.around(calculated[0:13], decimals=4)
        calculated[13:37] = np.around(calculated[13:37], decimals=5)
        calculated[37:63] = np.around(calculated[37:63], decimals=6)
        calculated[63:70] = np.around(calculated[63:70], decimals=7)
        calculated[70:78] = np.around(calculated[70:78], decimals=8)
        calculated[78:] = np.around(calculated[78:], decimals=9)
        
        #   Assign zero to the expected failing wavelengths.
        index_fail = [22, 70]
        reference[index_fail, 2] = 0
        calculated[index_fail] = 0
                
        #   Test if equal to reference.
        np.testing.assert_allclose(calculated, reference[:, 2], rtol=1e-4)
        
    @unittest.skip("Skip the expected failing wavelengths.")
    def test_failing_wavelengths(self):
        #   Specify the data file.
        filename_reference = os.path.join(data_base_path, "bucholtz_1995_table_2.csv")
        
        #   Load the reference data.
        reference = np.loadtxt(filename_reference, delimiter='\t', skiprows=1)
        
        #   Specify the failing wavelength index.
        index_fail = [22, 70]

        #   Calculate the total Rayleigh volume-scattering coefficient. (km^-1)
        calculated = volume_scattering_coefficient(reference[index_fail, 0], P_s, T_s)

        #   Round the values to the desired decimal.
        calculated[0] = np.around(calculated, decimals=5)
        calculated[1] = np.around(calculated, decimals=8)
        
        #   Test if equal to reference.
        np.testing.assert_allclose(calculated, reference[index_fail, 1], rtol=1e-4)
        
    def test_last_working_values(self):
        #   Specify the data file.
        filename_reference = os.path.join(data_base_path, "bucholtz_1995_last_working_values.csv")

        #   Load the data from the file.
        reference = np.loadtxt(filename_reference, delimiter='\t', skiprows=1)

        #   Calculate the total Rayleigh volume-scattering coefficient. (km^-1)
        calculated = volume_scattering_coefficient(reference[:, 0], P_s, T_s)

        # Test if equal to reference.
        np.testing.assert_allclose(calculated, reference[:, 2], rtol=1e-6)


if __name__ == "__main__":
    unittest.main()

bash_script.py
#!/bin/bash

#. /d/Anaconda/etc/profile.d/conda.sh    #here path to the conda prompt
#conda activate polly

python C:/Users/eliza/.PyCharmCE2019.3/config/scratches/scratch_24.py   #here path to the low_base script

for file in /d/march/height/*; do                   #here path to the created in low_base script height dir
name=${file##*/}
sed 's/\r//g' "$file"

cd /d/march/profiles/                               #here path to the created in low_base script profiles dir
while read -r line

#below after -d path to dir where converted files store
do
    #printf '%s\n' "$line"
	for i in $line; do
	echo $i
	polly2scc -p pollyxt_tropos -l Leipzig -t picasso -f $name -d C:/Users/eliza/Desktop/tropos/data/ --range_e 1000 $line --range_b 1000 $line --camp_info C:/Users/eliza/Desktop/polly2scc/config/Leipzig_campaign_info_5.toml; done
	python C:/Users/eliza/.PyCharmCE2019.3/config/scratches/scratch_25.py    # here path to the plot script


done < "$file"
done
for file in C:/Users/eliza/Desktop/tropos/data/*.nc; do rm $file; done       # here delete all files from the converted dir




plot_script.py
import matplotlib.pyplot as plt
from netCDF4 import Dataset
import os, fnmatch

import numpy as np
RAW_PATH= 'C:\\Users\\baars\\Desktop\\earlient_converter\\test_data\\' #path where the converted file are
Data_path = 'C:\\Users\\baars\\Desktop\\earlient_converter\\eliza_test_data\\out\\'
names = []
dict_back =  {}
dict_hei = {}
dict_ext = {}
dict_rat = {}
dict_dep = {}
dict_vol = {}





for name in fnmatch.filter(os.listdir(path=RAW_PATH), '*.nc'):
    names.append(name)
    backscatter = []
    extinction = []
    volume = []
    lidar_ratio = []
    depol = []
    lidar_ratio = []
    list_ext = []
    list_dep = []
    height = []

    file = Dataset(RAW_PATH + name)
    #height = file.variables["altitude"][:]

# next code is appending .nc variables to lists, also checking if some of vars exost as extinction, etc
    for i in range(file.dimensions["altitude"].size):

        backscatter.append(float(file.variables["backscatter"][0,0,i]))
        height.append(float(file.variables["altitude"][i]))
#        lidar_ratio.append(float(file.variables[""]))
        if "extinction" in file.variables:
            extinction.append(float(file.variables["extinction"][0, 0, i]))
            lidar_ratio.append(float(file.variables["extinction"][0,0,i])/float(file.variables["backscatter"][0, 0, i]))
        if "particledepolarization" in file.variables:
            depol.append(float(file.variables["particledepolarization"][0, 0, i]))
        if "volumedepolarization" in file.variables:
            volume.append(float(file.variables["volumedepolarization"][0, 0, i]))


    # create dictionaries for plotting
    dict_2 = {name.split("_")[-1] : tuple(backscatter)}
    dict_back.update(dict_2)

    dict_4 = {name.split("_")[-1]: tuple(height)}
    dict_hei.update(dict_4)

    dict_1 = {name.split("_")[-1]: tuple(extinction)}
    dict_ext.update(dict_1)

    dict_5 = {name.split("_")[-1]: tuple(lidar_ratio)}
    dict_rat.update(dict_5)

    dict_3 = {name.split("_")[-1]: tuple(depol)}
    dict_dep.update(dict_3)

    dict_6 = {name.split("_")[-1]: tuple(volume)}
    dict_vol.update(dict_6)






# plotting routine from dictionaries

fig, (ax1, ax2, ax3, ax4) = plt.subplots(1, 4)
fig.suptitle('Horizontally stacked subplots')


ax1.set_title("Backscatter, 1/(m*sr)")

ax1.plot(dict_back.get("b532.nc"), dict_hei.get("b532.nc"),label="b532", color='red', linestyle='--', linewidth='2')
ax1.plot(dict_back.get("b1064.nc"), dict_hei.get("b1064.nc"), label="b1064",linestyle=':', linewidth='2')
ax1.plot(dict_back.get("b355.nc"), dict_hei.get("b355.nc"), label="b355", linestyle='--', linewidth='2')
ax1.plot(dict_back.get("e355.nc"), dict_hei.get("e355.nc"), label="e355", linestyle=':', linewidth='1')
ax1.plot(dict_back.get("e532.nc"), dict_hei.get("e532.nc"),  label="e532", color='blue',linestyle=':', linewidth='1')
ax1.set_xlim(left=0)
#ax1.set_xticks(rotation=45)
ax1.legend()



ax2.set_title("Extinction, 1/m")

ax2.plot(dict_ext.get("e355.nc"), dict_hei.get("e355.nc"), label="e355")
ax2.plot(dict_ext.get("e532.nc"), dict_hei.get("e532.nc"), label="e532" )
ax2.set_xlim(left=0)
ax2.legend()




for i in dict_dep:
    if len(dict_dep.get(i)) != 0 and all(dict_dep.get(i))!="nan":
        ax3.plot(dict_dep.get(i), dict_hei.get(i), label="particledepol" + "_" +  i)
        ax3.plot(dict_vol.get(i), dict_hei.get(i), label="volumedepol" +"_" + i)
        ax3.set_title("Particle/Volume depolarization")
        #ax3.set_xlim(0,0.5)
ax3.legend()

for i in dict_rat:
    if len(dict_rat.get(i)) != 0 and all(dict_rat.get(i))!="nan":
        ax4.plot(dict_rat.get(i), dict_hei.get(i), label=i)
        ax4.set_title("Lidar Ratio")
        ax4.set_xlim(0,250)
        #ax4.set_xticks(rotation=45)
ax4.legend()

fig = plt.gcf()
#plt.gca().set_xlim(left=0)
fig.autofmt_xdate(rotation=45)
fig.set_size_inches((15, 11), forward=False)
fig.savefig(Data_path + name +'.png', dpi=500)

plt.close()




low_base_script.py
import numpy as np
import os, fnmatch

import datetime
import math
from netCDF4 import Dataset
import shutil

name = 0
dict_pro = {}
dict_cloud = {}
RAW_PATH = 'C:\\Users\\baars\\Desktop\\earlient_converter\\eliza_test_data\\new_test_data\\'          # path where you want to store files: profiles, cloudinfo, etc
Data_path = 'C:\\Users\\baars\\Desktop\\earlient_converter\\eliza_test_data\\out\\' # path for the converted files
first_date = datetime.datetime(1970, 1, 1)
names = []
value_no_cloud = 15000

# Create directories to separate different files profiles, cloudinfo, etc
dirName = 'cloud'
try:
    # Create target Directory
    os.mkdir(RAW_PATH + dirName)
    print("Directory " , dirName ,  " Created ")
except FileExistsError:
    print("Directory " , dirName ,  " already exists")

# Create directory
dirName = 'height'
try:
    # Create target Directory
    os.mkdir(RAW_PATH + dirName)
    print("Directory " , dirName ,  " Created ")
except FileExistsError:
    print("Directory " , dirName ,  " already exists")

# Create directory
dirName = 'profiles'
try:
    # Create target Directory
    os.mkdir(RAW_PATH + dirName)
    print("Directory " , dirName ,  " Created ")
except FileExistsError:
    print("Directory " , dirName ,  " already exists")

dirName = 'plots'
try:
    # Create target Directory
    os.mkdir(RAW_PATH + dirName)
    print("Directory " , dirName ,  " Created ")
except FileExistsError:
    print("Directory " , dirName ,  " already exists")
#for file in os.listdir(RAW_PATH + "profiles/"):
 #   try:
  #      s = Dataset(RAW_PATH + "profiles/" + file)
   # except:
    #    continue
    #if "aerBsc_raman_1064" in s.variables or "aerBsc_raman_355" in s.variables or "aerBsc_raman_532" in s.variables :
     #   g = np.array(s.variables["aerBsc_raman_1064"])
      #  print(g)
       # print(np.all([g,np.nan]))
        #if s.variables['aerBsc_raman_532'][:]

        #shutil.copy(RAW_PATH + "profiles/" + file, Data_path  + "data_w_back/")

# copy files from raw-path dir for separation
for file in fnmatch.filter(os.listdir(path=RAW_PATH), '*_cloudinfo.nc'):
    shutil.copy(RAW_PATH + file, RAW_PATH + "cloud\\")
for file in fnmatch.filter(os.listdir(path=RAW_PATH), '*_profiles.nc'):
    if "NR" not in file and "OC"  not in file:
        shutil.copy(RAW_PATH + file, RAW_PATH + "profiles\\")


#creating dictionaries for further comparison between profiles  and cloudinfo files
for name in os.listdir(RAW_PATH + 'profiles\\'):
    #print(name)

    list_names_prof = ["Year", "Month", "Day", "Day_Week", "Location", "a", "b", "c", "Start_time", "End_time", "ext"]


    lst = name.split(sep="_")
    lst2 = name.split(sep="_")[:6]
    for j in range(len(lst)):
        dict1 = {list_names_prof[j]: lst[j]}
        dict_pro.update(dict1)
    for name1 in os.listdir(RAW_PATH + 'cloud\\'):
        d = name1.split(sep="_")[:6]
        #print(d, lst2)
        cloud = name1.split(sep="_")
        if lst2 == d:

            list_names_cloud = ["Year", "Month", "Day", "Day_Week", "Location", "a", "b", "c", "ext"]


            for i in range(len(cloud)):
                dict2 = {list_names_cloud[i]: cloud[i]}
                dict_cloud.update(dict2)

            time_since = datetime.datetime(int(dict_cloud.get("Year")), int(dict_cloud.get("Month")),
                                       int(dict_cloud.get("Day"))) - first_date
            seconds = int(time_since.total_seconds())


            r = Dataset(RAW_PATH + "cloud\\" + name1)
            d = Dataset(RAW_PATH + "profiles\\" + name)
            #print(dict_pro)
            #print(dict_cloud)
           # print("MATCH!", name1)
            low_base = []
            st_time = r.variables['time'][0]
            hour_start = int((st_time - seconds) / 3600)
            min_start = int((st_time - seconds - hour_start * 3600) / 60)
            sec_start = int(st_time - seconds -  hour_start * 3600 - min_start * 60)

            #low base searching
            for i in range(r.dimensions["time"].size):
                #print(int((r.variables["time"][i] - seconds)/3600), int(dict_pro.get("Start_time")[:2]))
                if int((r.variables["time"][i] - seconds)/3600) == int(dict_pro.get("Start_time")[:2]):

                    if math.isnan(r.variables["cloud_base_height"][i][0]):
                        continue
                    else:
                        low_base.append(int(r.variables["cloud_base_height"][i][0]))
            #print(name.split(sep='cloudinfo')[0])

            for i in low_base:
                if i < 1000:
                    low_base.remove(i)
            #print(low_base)

            if len(low_base) == 0:


                break
                #outF.write(str(d.dimensions["height"].size))
                #outF.write("\n")
            else:                     # saving low base values for different text files for bash script
                #print(min(low_base))
                outF = open(
                    Data_path + "height\\ " + name1.split(sep='cloudinfo')[0] + str(dict_pro.get("Start_time")) + "_" + str(
                        dict_pro.get("End_time")) + "_" + "profiles" + ".nc", "w")
                outF.write(str(min(low_base)))
                outF.write("\n")
                outF.close()

